providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -0,0 +1,590 @@
+{
+    if (key == NULL)
+        return 0;
+
+    if (key->keytype != MLKEM_KEY_TYPE_768)
+        return 0;
+
+    if (!ossl_param_build_set_octet_string(tmpl, params,
+                                           OSSL_PKEY_PARAM_PUB_KEY,
+                                           key->pubkey, MLKEM768_PUBLICKEYBYTES))
+        return 0;
+
+    if (include_private
+        && key->seckey != NULL
+        && !ossl_param_build_set_octet_string(tmpl, params,
+                                              OSSL_PKEY_PARAM_PRIV_KEY,
+                                              key->seckey, MLKEM768_SECRETKEYBYTES))
+        return 0;
+
+    return 1;
+}
+
+static int mlkem_export(void *key, int selection, OSSL_CALLBACK *param_cb,
+                        void *cbarg)
+{
+    MLKEM768_KEY *mkey = key;
+    OSSL_PARAM_BLD *tmpl;
+    OSSL_PARAM *params = NULL;
+    int ret = 0;
+
+    debug_print("MLKEMKM export %p\n", key);
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    tmpl = OSSL_PARAM_BLD_new();
+    if (tmpl == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        int include_private = ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0);
+
+        if (!key_to_params(mkey, tmpl, NULL, include_private))
+            goto err;
+    }
+
+    params = OSSL_PARAM_BLD_to_param(tmpl);
+    if (params == NULL)
+        goto err;
+
+    ret = param_cb(params, cbarg);
+    OSSL_PARAM_free(params);
+err:
+    OSSL_PARAM_BLD_free(tmpl);
+    debug_print("MLKEMKM export result %d\n", ret);
+    return ret;
+}
+
+static int ossl_mlkem_key_fromdata(MLKEM768_KEY *key,
+                                   const OSSL_PARAM params[],
+                                   int include_private)
+{
+    size_t privkeylen = 0, pubkeylen = 0;
+    const OSSL_PARAM *param_priv_key = NULL, *param_pub_key;
+    unsigned char *pubkey;
+
+    if (key == NULL)
+        return 0;
+
+    if (key->keytype != MLKEM_KEY_TYPE_768)
+        return 0;
+
+    param_pub_key = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (include_private)
+        param_priv_key =
+            OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PRIV_KEY);
+
+    if (param_pub_key == NULL && param_priv_key == NULL)
+        return 0;
+
+    if (param_priv_key != NULL) {
+        if (!OSSL_PARAM_get_octet_string(param_priv_key,
+                                         (void **)&key->seckey,
+                                         MLKEM768_SECRETKEYBYTES,
+                                         &privkeylen))
+            return 0;
+        if (privkeylen != MLKEM768_SECRETKEYBYTES) {
+            debug_print("sec key len mismatch in import: %ld vs %d: HOWCAN?\n",
+                        privkeylen, MLKEM768_SECRETKEYBYTES);
+            OPENSSL_secure_clear_free(key->seckey, privkeylen);
```

Comment:
This looks suspect? `key->seckey` doesn't point to an independently allocated buffer of bytes (we never call `OPENSSL_secure_malloc()` to allocate it independently). So calling `OPENSSL_secure_clear_free` on it (and then setting it to NULL int the line below) looks incorrect to me?

Reply:
I kind of read the documentation of `OSSL_PARAM_get_octet_string` to do this allocation, no?

Reply:
And so (wrong) is the logic around pubkey (even more, as it's not the pubkey but the encoded_pubkey that's restored this way). This code clearly has never run (as per #ifdef UNDEF it's wrapped into). This needs changing to (calling/using) proper marshaling functions. This now begs the question: Do we remove this wrong code completely to add it back as and when it's needed or "move forward" work on https://github.com/openssl/openssl/issues/25885?